optaplanner-core/src/main/java/org/optaplanner/core/impl/io/jaxb/JaxbIO.java

```diff
@@ -52,25 +79,125 @@ public JaxbIO(Class<T> rootClass, int indentation) {
         this.rootClass = rootClass;
         this.indentation = indentation;
         try {
-            JAXBContext jaxbContext = JAXBContext.newInstance(rootClass);
-            unmarshaller = jaxbContext.createUnmarshaller();
+            jaxbContext = JAXBContext.newInstance(rootClass);
             marshaller = jaxbContext.createMarshaller();
             marshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, true);
             marshaller.setProperty(Marshaller.JAXB_ENCODING, StandardCharsets.UTF_8.toString());
         } catch (JAXBException jaxbException) {
-            String errMessage = String.format("Unable to create JAXB marshaller or unmarshaller for a root element class %s.",
+            String errMessage = String.format("Unable to create JAXB Marshaller for a root element class (%s).",
                     rootClass.getName());
-            throw new IllegalStateException(errMessage, jaxbException);
+            throw new OptaPlannerXmlSerializationException(errMessage, jaxbException);
         }
     }
 
     public T read(Reader reader) {
         Objects.requireNonNull(reader);
         try {
-            return (T) unmarshaller.unmarshal(reader);
+            @SuppressWarnings("unchecked")
+            final T object = (T) createUnmarshaller().unmarshal(reader);
+            return object;
         } catch (JAXBException jaxbException) {
-            String errMessage = String.format("Unable to read the %s from XML.", rootClass.getName());
-            throw new XmlUnmarshallingException(errMessage, jaxbException);
+            String errMessage = String.format(ERR_MSG_READ, rootClass.getName());
+            throw new OptaPlannerXmlSerializationException(errMessage, jaxbException);
+        }
+    }
+
+    public T readAndValidate(Reader reader, String schemaResource) {
+        Objects.requireNonNull(reader);
+        Objects.requireNonNull(schemaResource);
```

Comment:
Do you consider it more readable? I was thinking the opposite: doing the check on line 111 adds one more nested call.

Reply:
Somehow I don't like that this code relies on a side-effect of `requireNonNull(...)` - it will throw an exception when something's wrong. While this:

`String nonNullSchemaResource = Objects.requireNonNull(schemaResource);`

It still relies on the same side effect, but it's doesn't hurt my eyes so much. Plus if you accidentally delete this line, the code will no longer compile - if you delete the original line, the code will have just silently lost a feature.